[hotfix/UT] Prevent cache corruption when vLLM prefix cache overlaps with LMCache

DongDongJu opened the pull request:
Updates:
After discussion with @ApostaC, we currently decided to set the default value of `save_unfull_chunk` as false.
I changed the code, doc and modify the ut cover for both case(True and False).

# history 
During investigation https://github.com/LMCache/LMCache/issues/2136 with local disk backend test,
but I found out that when vLLM prefix cache and LMCache align to the same chunk boundary, LMCache loads tokens already cached by vLLM, causing cache corruption.

**Example:** `vllm_cached_tokens=14816`, `lmcache_cached_tokens=14820` → both round to `14592` → previous logic overwrites vLLM cache

**What this PR does / why we need it**:
Fixes cache corruption when vLLM prefix cache and LMCache align to the same chunk boundary. 
Previously, LMCache would load tokens already cached by vLLM, overwriting vLLM's cache and causing incorrect outputs.

This PR skips LMCache retrieval when all chunk-aligned tokens are cached by vLLM, letting vLLM recompute partial chunks to avoid data inconsistency.

**Special notes for your reviewers**:
- Skip logic: `if retrieve_end_aligned <= masked_token_count: continue`
- When skipped, vLLM recomputes tokens from `vllm_cached_tokens` to `lmcache_cached_tokens`, which may cause slightly different output due to floating-point non-determinism, but prevents cache corruption
- Tested with same prompt four times (cold → warm), `temperature=0`, `top_k=1`

Compared to https://github.com/LMCache/LMCache/issues/2206's result.
we can get below result with 4 times same query.
<img width="945" height="630" alt="image" src="https://github.com/user-attachments/assets/417c0ecb-cbf7-4dc6-928c-adc879658628" />
<img width="937" height="631" alt="image" src="https://github.com/user-attachments/assets/b5e543d2-7187-4632-b067-4e5aede84ce5" />

**If applicable**:
- [ ] this PR contains user facing changes - docs added
- [x] this PR contains unit tests

DongDongJu commented:
I ran w/o layerwise cuz it has issue now.

DongDongJu committed:
Change save_unfull_chunk default to False and update documentation

DongDongJu commented:
After discussing with @ApostaC offline, currently we set save_unfull_chunk to false as a default variable.
result in same request return same result
```
<think>
Okay, let's see. The user wants a summary of bash in two sentences. First, I need to recall the main features of bash. It's a shell interpreter, supports various options, handles commands, and has some advanced features like the --debugger and --version. Also, it's a POSIX-compliant shell. 
Wait, the user provided a detailed description, but they want it in two sentences. So I should condense that into two concise points. Maybe start with its role as a shell interpreter and mention key features like options, command execution, and compatibility. That should cover it.
</think>
Bash is a powerful, POSIX-compliant shell interpreter that executes commands from standard input or files, supports various options, and handles advanced features like interactive mode and restricted shells.
Cold TTFT: 33.647 seconds
Querying vLLM server with warm LMCache Disk Offload
<think>
Okay, let's see. The user wants a summary of bash in two sentences. First, I need to recall the main features of bash. It's a shell interpreter, supports various options, handles commands, and has some advanced features like the --debugger and --version. Also, it's a POSIX-compliant shell. 
Wait, the user provided a detailed description, but they want it in two sentences. So I should condense that into two concise points. Maybe start with its role as a shell interpreter and mention key features like options, command execution, and compatibility. That should cover it.
</think>

Bash is a powerful, POSIX-compliant shell interpreter that executes commands from standard input or files, supports various options, and handles advanced features like interactive mode and restricted shells.
Warm TTFT: 0.147 seconds

TTFT Improvement: 33.500 seconds     (229.3x faster)
```

DongDongJu committed:
Merge branch 'dev' into devel

DongDongJu commented:
@ApostaC @sammshen Could you take a look?

DongDongJu committed:
Merge branch 'dev' into devel

DongDongJu committed:
Add save_unfull_chunk parametrization to benchmark tests

- Add parametrization for save_unfull_chunk (False/True) to all benchmark tests
- Update test assertions to handle both chunk-aligned and full token counts
- Fix line length style issues

Signed-off-by: Dongjoo Seo <dongjoo.seo1@samsung.com>
Signed-off-by: DongDongJu <commisori28@gmail.com>

DongDongJu committed:
test: Refactor get_expected_count and fix test_paged_store_kv_tensors_mask for save_unfull_chunk

- Add module-level get_expected_count helper function to avoid duplication
- Parametrize all test_paged_* functions with save_unfull_chunk=[False, True]
- Fix test_paged_store_kv_tensors_mask to handle chunk-aligned mask requirements
  when save_unfull_chunk=True
- Maintain original test flow while supporting both save_unfull_chunk values
- Update all get_expected_count calls to use module-level function

Signed-off-by: Dongjoo Seo <dongjoo.seo1@samsung.com>
Signed-off-by: DongDongJu <commisori28@gmail.com>

DongDongJu committed:
Merge branch 'dev' into devel

DongDongJu commented:
/gemini Please review the PR

DongDongJu committed:
test: Refactor get_expected_count in test_benchmark.py to module level

- Add module-level get_expected_count helper function to avoid duplication
- Remove duplicate get_expected_count definitions from test functions
- Update all get_expected_count calls to use module-level function with
  save_unfull_chunk and chunk_size parameters
- Improve code maintainability and consistency with test_cache_engine.py

Signed-off-by: Dongjoo Seo <dongjoo.seo1@samsung.com>
Signed-off-by: DongDongJu <commisori28@gmail.com>

DongDongJu committed:
Merge branch 'devel' of https://github.com/DongDongJu/LMCache into devel

DongDongJu commented on tests/benchmarks/test_benchmark.py:
```diff
@@ -152,6 +162,14 @@ def test_store_1GB(benchmark, backend, create_config, autorelease_v1):
+            return token_len
+        else:
+            # Round down to chunk boundary
+            return (token_len // chunk_size) * chunk_size
```
You are right. I resolved it by https://github.com/LMCache/LMCache/pull/2207/commits/bcb86070573b4f72b3f526b5fd3f71028d80daa9

DongDongJu commented:
@sammshen @ApostaC ready to review it. I just made `save_unfull_chunk` to false and modify the docs and ut tests based on that changes. Tried to keep the original ut behavior.

ApostaC reviewed (approved):
LGTM! Thanks!

@sammshen @maobaolong @KuntaiDu Can one of you take a quick review at this PR? We are going to make `save_unfull_chunk` default to false for a while until we've got a clear sense about the potential correctness issue. Thanks!

DongDongJu committed:
Merge branch 'dev' into devel

DongDongJu committed:
Merge branch 'dev' into devel

YaoJiayi reviewed (commented):
@DongDongJu thanks! lgtm in general. Just one question: will pd backend be affected by this?

DongDongJu committed:
config: Auto-enable save_unfull_chunk for PD mode

PD (Peer-to-Peer Disaggregation) requires save_unfull_chunk=True to
ensure complete KV cache transfer from prefill node to decode node.
Without this, partial chunks would be discarded, causing incomplete
KV cache transfer and wrong results on the decode node.

- Automatically set save_unfull_chunk=True when enable_pd=True
- Add warning log when auto-setting save_unfull_chunk
- Add info log when save_unfull_chunk is already True in PD mode
- Ensure PD users get correct configuration without manual intervention

Signed-off-by: Dongjoo Seo <dongjoo.seo1@samsung.com>
Signed-off-by: DongDongJu <commisori28@gmail.com>

DongDongJu commented:
> @DongDongJu thanks! lgtm in general. Just one question: will pd backend be affected by this?

You are right. I added https://github.com/LMCache/LMCache/pull/2207/commits/25310af6351c70bb34fa1d88c4ae82e7690f550a to force sending full cache if pd is enabled.

YaoJiayi reviewed (approved):
Thanks for this quick fix!

DongDongJu committed:
Auto-enable save_unfull_chunk=True for MLA mode

MLA (Multi-Level Attention) mode requires save_unfull_chunk=True for
correct KV cache storage and retrieval. Without this, partial chunks
would be discarded, causing incomplete cache and incorrect results.

This change automatically sets save_unfull_chunk=True when MLA mode is
detected during initialization, similar to how PD mode handles this.

Also update local_cpu_mla.yaml to set discard_partial_chunks=false
to align with the auto-enabled save_unfull_chunk=True behavior.

Signed-off-by: Dongjoo Seo <dongjoo.seo1@samsung.com>
Signed-off-by: DongDongJu <commisori28@gmail.com>

DongDongJu committed:
Merge branch 'dev' into devel

DongDongJu committed:
Fix Ruff F841: Remove unused use_mla variable

Remove unused use_mla variable that was causing Ruff F841 error.
The variable was left over from previous MLA-specific handling that
has been moved to _init_lmcache_engine.

Signed-off-by: Dongjoo Seo <dongjoo.seo1@samsung.com>
Signed-off-by: DongDongJu <commisori28@gmail.com>

DongDongJu commented:
I think there are some bugs related chunked prefill with mla mode in vllm https://github.com/vllm-project/vllm/issues/26042.
I will take a look after make next PR.

yoo-kumaneko commented on lmcache/integration/vllm/vllm_v1_adapter.py:
```diff
@@ -501,6 +501,22 @@ def _init_lmcache_engine(
     ):
         raise ValueError("MLA only works with naive serde mode..")
 
+    # MLA requires save_unfull_chunk=True for correct KV cache storage and retrieval.
```
@DongDongJu I might be missing something. Could you explain why save_unfull_chunk affects the storage of the MLA cache? Also, does MLA refers to Multi-head Latent Attention or Multi Level Attention?

DongDongJu replied:
Hello @yoo-kumaneko, Happy new year.
Sry mb, it was Deepseek's Multi-head Latent Attention.
I will change it soon.
I simply thought that it was buggy(https://github.com/vllm-project/vllm/issues/26042) before.
So maybe we can remove the entire block for now.
what do you think?

yoo-kumaneko replied:
> Hello @yoo-kumaneko, Happy new year. Sry mb, it was Deepseek's Multi-head Latent Attention. I will change it soon. I simply thought that it was buggy([vllm-project/vllm#26042](https://github.com/vllm-project/vllm/issues/26042)) before. So maybe we can remove the entire block for now. what do you think?

Happy New Year. 
Yes, we can remove the entire block. I don't think it has any effect on the cache storage for MLA now.

DongDongJu replied:
Let me make the hotfix for this.

DongDongJu replied:
made PR for changes https://github.com/LMCache/LMCache/pull/2349
